### ISSUE TITLE
Project Status & Roadmap: From Knowledge Graphs to Graph-Enhanced Agent Skills

### ISSUE THREAD

**Issue by rysweet** [rysweet/agent-kgpacks#141]
### Project Overview

WikiGR has evolved from a Wikipedia knowledge graph builder into a **complete platform for creating reusable, graph-enhanced agent skills (Knowledge Packs)** that provably surpass an agent's training data and web search on specialized domains.

#### Core Knowledge Graph System (Production)
- **31,777 Wikipedia articles** indexed with 87,500 entities and 54,393 relationships
- **6.2M article links** in a 3.0 GB Kuzu embedded graph database
- **Hybrid retrieval**: vector search + graph traversal + keyword matching
- **KG Agent**: Cypher-based query engine with semantic search, 23/23 gold standard tests passing
- **Full-stack PWA**: React TypeScript frontend + FastAPI backend with SSE chat streaming

#### Web Content Source (PR #138 — Merged)
Six gaps closed to achieve full feature parity between web and Wikipedia sources:

| Feature | Status | Details |
|---------|--------|---------|
| Unified schema | ✅ | CLI calls `ryugraph_schema.create_schema()` with Entity, Fact, Chunk tables |
| Shared processor | ✅ | `ArticleProcessor` accepts `ContentSource` protocol — no code duplication |
| Chunk creation | ✅ | Automatic via processor pipeline for both sources |
| LLM extraction | ✅ | Optional, gated on `ANTHROPIC_API_KEY`; extracts entities, relationships, facts |
| BFS link expansion | ✅ | `--max-depth`, `--max-links` flags; visited set, rate limiting, same-domain filter |
| Update command | ✅ | `wikigr update --source=web --urls=file.txt` with duplicate detection |

**Security hardening** (4 CRITICAL fixes):
- DNS rebinding SSRF mitigation (TOCTOU prevention)
- IPv6 site-local address blocking (fec0::/10)
- API key sanitization in log output
- Unicode homoglyph protection (IDNA normalization)

**Tests**: 87 passing (13.82% coverage), including 41 dedicated security tests.

#### Knowledge Packs System (PR #140 — Merged)
All five implementation phases complete:

| Phase | Deliverable | Status | Tests |
|-------|-------------|--------|-------|
| **1. Pack Format** | Manifest dataclass, validator, parser | ✅ | 32 |
| **2. Skills Integration** | Auto-discovery, skill template generator, registry | ✅ | 43 |
| **3. Evaluation Framework** | 3-baseline comparison, metrics, runner | ✅ | 44 |
| **4. Distribution** | tar.gz packaging, installer, versioning, registry API | ✅ | 56 |
| **5. CLI Commands** | 8 subcommands: create, install, list, info, eval, update, remove, validate | ✅ | 16 |

**Total**: 175 pack tests passing, plus 87 bootstrap tests = **262 tests** across the project.

### Pack Directory Structure
```
~/.wikigr/packs/physics-expert/
├── manifest.json          # Metadata, graph stats, eval scores
├── pack.db/               # Kuzu knowledge graph database
├── skill.md               # Claude Code skill definition (auto-generated)
├── kg_config.json         # KG Agent retrieval configuration
├── eval_questions.jsonl   # Evaluation benchmark questions (optional)
└── README.md              # Pack documentation (optional)
```

### CLI Commands Available Today
```bash
# Knowledge graph construction
wikigr create --source=web --urls=urls.txt --max-depth=2 --db=output.db
wikigr create --topics=topics.md --target=5000 --db=output.db
wikigr update --source=web --urls=more.txt --db=existing.db
wikigr status --db=output.db

# Knowledge pack lifecycle
wikigr pack create --name=physics-expert --topics=topics.txt --target=5000
wikigr pack install physics-expert.tar.gz
wikigr pack list [--format=json]
wikigr pack info physics-expert [--show-eval-scores]
wikigr pack eval physics-expert [--questions=custom.jsonl] [--save-results]
wikigr pack update physics-expert --from=v1.1.0.tar.gz
wikigr pack remove physics-expert [--force]
wikigr pack validate ~/.wikigr/packs/physics-expert [--strict]
```

### Architecture Summary
```
Content Sources (Web, Wikipedia)
         │
         ▼
ArticleProcessor (ContentSource protocol)
  ├── Section embedding (paraphrase-MiniLM-L3-v2, 384-dim)
  ├── Chunk creation (for hybrid retrieval)
  ├── LLM extraction (entities, relationships, facts — optional)
  └── Category/link insertion
         │
         ▼
Kuzu Graph Database
  ├── Article → Section → Chunk (HAS_SECTION, HAS_CHUNK)
  ├── Section → Entity (HAS_ENTITY)
  ├── Section → Fact (HAS_FACT)
  ├── Entity → Entity (ENTITY_RELATION)
  └── HNSW vector index on Section.embedding
         │
         ▼
Knowledge Pack Bundle
  ├── manifest.json + pack.db + skill.md + kg_config.json
  ├── Evaluation framework (3 baselines)
  └── Distribution (.tar.gz, registry)
         │
         ▼
Claude Code Skill (auto-discovered)
  └── KG Agent hybrid retrieval → enriched agent responses
```

### What's NOT Done Yet

The following items represent the gap between "system works" and "system delivers measurable value with real packs":

1. **No real knowledge packs exist yet** — the system is built but no actual packs have been created, evaluated, or distributed
2. **Evaluation baselines use mocked API calls** — the Anthropic SDK integration in baselines.py returns placeholder responses; real Claude API evaluation has not been run
3. **`pack create` builds manifest but doesn't run the full pipeline** — it doesn't invoke `wikigr create` to actually build the graph, then package it
4. **KG Agent is not wired into pack evaluation** — the `KnowledgePackEvaluator` has a stub `_retrieve_context()` that doesn't call the actual KG Agent
5. **Skill auto-load not tested end-to-end** — skill.md files are generated correctly, but loading them as actual Claude Code skills hasn't been validated
6. **Registry API is mock-only** — `PackRegistryClient` has the HTTP interface but no backend service
7. **No CI job runs pack tests** — the 175 pack tests run locally but aren't included in `ci.yml`
8. **Cleanup needed** — temporal docs (`PHASE4_DISTRIBUTION.md`), example pack artifacts, feature branches

---

### Priority 1: Create First Real Knowledge Pack (Proof of Value)

**Goal**: Build and evaluate a real physics-expert pack that demonstrably surpasses Claude's training data on physics questions.

**Tasks**:

- [ ] **1a. Create physics evaluation question set** — Write 50-100 physics questions with ground truth answers spanning quantum mechanics, thermodynamics, relativity, classical mechanics. Include difficulty levels (easy/medium/hard). Save as `eval/physics/questions.jsonl`.

- [ ] **1b. Create physics topics file** — Curate 200-500 Wikipedia seed topics for physics domain. Save as `eval/physics/topics.txt`.

- [ ] **1c. Build the graph** — Run `wikigr create --topics=eval/physics/topics.txt --target=5000 --db=packs/physics-expert/pack.db` to build a real 5K-article physics knowledge graph with LLM extraction enabled.

- [ ] **1d. Wire KnowledgePackEvaluator to real KG Agent** — Replace `_retrieve_context()` stub in `wikigr/packs/eval/baselines.py` with actual KG Agent hybrid retrieval call against `pack.db`.

- [ ] **1e. Wire baseline evaluators to real Anthropic API** — Update `TrainingBaselineEvaluator` and `WebSearchBaselineEvaluator` in `baselines.py` to make real `anthropic.Anthropic().messages.create()` calls instead of returning mocked responses.

- [ ] **1f. Run full 3-baseline evaluation** — Execute `wikigr pack eval physics-expert --save-results` with real API calls. Compare training vs web search vs knowledge pack on all 50-100 questions. Document actual accuracy, hallucination rate, and citation quality.

- [ ] **1g. Package and validate** — Run `wikigr pack validate physics-expert --strict` and create distributable archive.

**Acceptance criteria**: A real pack exists with measured eval scores showing pack surpasses at least one baseline on at least one metric.

---

### Priority 2: End-to-End `pack create` Pipeline

**Goal**: `wikigr pack create` runs the full pipeline (graph build + extraction + packaging) in a single command.

**Tasks**:

- [ ] **2a. Connect `pack create` to `wikigr create`** — `cmd_pack_create()` in cli.py currently generates manifest and scaffold but doesn't build the graph. Wire it to call `_expand_seeds()` or `_create_from_urls()` depending on `--source`.

- [ ] **2b. Auto-run LLM extraction after graph build** — After sections are loaded, automatically run entity/relationship extraction if `ANTHROPIC_API_KEY` is set.

- [ ] **2c. Auto-generate eval questions** — Optionally use LLM to generate evaluation questions from the graph content (e.g., "Based on the entities in this graph, generate 50 domain-specific questions with ground truth answers").

- [ ] **2d. Compute graph stats automatically** — After build, query the database for article count, entity count, relationship count, and DB size, then populate `manifest.json` automatically.

- [ ] **2e. Integration test** — Test `wikigr pack create --name=test-pack --topics=test-topics.txt --target=10` end-to-end in CI.

---

### Priority 3: Wire KG Agent into Pack Skills

**Goal**: When a user invokes a pack skill, the KG Agent retrieves from the pack's graph database.

**Tasks**:

- [ ] **3a. KG Agent accepts arbitrary db_path** — Currently `KGAgent` in `wikigr/agent/kg_agent.py` is initialized with a specific database. Add constructor parameter for pack database path.

- [ ] **3b. Pack skill invocation loads KG Agent** — When Claude Code loads a pack's `skill.md`, the skill should instantiate a `KGAgent(db_path=pack.path / "pack.db")` and use it for retrieval.

- [ ] **3c. Skill.md template includes retrieval instructions** — Update `generate_skill_md()` to include executable instructions for Claude to call the KG Agent.

- [ ] **3d. End-to-end validation** — Test in a Claude Code session: install pack → invoke skill → verify KG Agent retrieves from pack graph → verify enriched response.

---

### Priority 4: Add Pack Tests to CI

**Goal**: Pack tests run in CI alongside existing tests.

**Tasks**:

- [ ] **4a. Add pack test job to ci.yml** — Add a job that runs `pytest tests/packs/ -v --timeout=60` across Python 3.10/3.11/3.12.

- [ ] **4b. Add CLI pack test to CI** — Run `pytest tests/cli/test_pack_commands.py -v`.

- [ ] **4c. Add coverage for wikigr/packs/** — Configure coverage to include `wikigr/packs/` in coverage reporting.

- [ ] **4d. Fix any CI-only failures** — Ensure pack tests pass in GitHub Actions environment (may need to handle missing `~/.wikigr/packs/` directory, no API keys, etc.).

---

### Priority 5: Build Second Pack (Microsoft Fabric GraphQL Expert)

**Goal**: Validate the system works for web-sourced (non-Wikipedia) content.

**Tasks**:

- [ ] **5a. Curate Microsoft Fabric GraphQL URLs** — Collect 100-200 URLs from `learn.microsoft.com/en-us/fabric/` covering Graph QL, data warehouse, lakehouse, real-time intelligence.

- [ ] **5b. Create evaluation questions** — Write 30-50 Fabric-specific questions (e.g., "How do you create a GraphQL resolver in Fabric?", "What's the difference between Direct Lake and import mode?").

- [ ] **5c. Build web-sourced pack** — Run `wikigr pack create --name=fabric-graphql-expert --source=web --urls=fabric-urls.txt --target=200`.

- [ ] **5d. Run evaluation** — Compare training vs web search vs pack on Fabric-specific questions.

- [ ] **5e. Document results** — Write up the comparison showing whether web-sourced packs can achieve the same quality as Wikipedia-sourced packs.

---

### Priority 6: Repository Cleanup

**Goal**: Clean up artifacts from the implementation sprint.

**Tasks**:

- [ ] **6a. Remove temporal docs** — Move `PHASE4_DISTRIBUTION.md`, `PHASE2_IMPLEMENTATION.md`, and similar implementation notes out of the repo or into `~/.amplihack/.claude/runtime/logs/`.

- [ ] **6b. Remove `SECURITY_FIXES.md` from repo root** — This was a sprint artifact; security details belong in commit messages and PR description, not a root-level file.

- [ ] **6c. Clean up feature branches** — Delete merged branches: `feat/issue-137-web-source-parity`, `feat/issue-139-knowledge-packs-design`, and any other stale branches.

- [ ] **6d. Remove worktrees** — Clean up `./worktrees/` directory.

- [ ] **6e. Update README.md** — Add Knowledge Packs section to the project README with usage examples and link to design doc.

- [ ] **6f. Update pyproject.toml** — Add `wikigr.packs` and dependencies (anthropic SDK) to the project configuration.

---

### Priority 7: Pack Registry Service (Future)

**Goal**: Enable publishing and sharing packs via a hosted registry.

**Tasks**:

- [ ] **7a. Design registry API** — Define REST endpoints: `POST /packs`, `GET /packs`, `GET /packs/{name}`, `GET /packs/{name}/download`.

- [ ] **7b. Implement registry backend** — Simple FastAPI service with file storage (S3 or Azure Blob) for pack archives.

- [ ] **7c. Wire `PackRegistryClient`** — Connect the existing mock client to the real registry.

- [ ] **7d. Add `wikigr pack publish`** — CLI command to publish pack to registry.

- [ ] **7e. Add `wikigr pack search`** — CLI command to search published packs.

---

### Priority 8: Advanced Features (Future)

- [ ] **8a. Multi-pack queries** — Query across multiple installed packs simultaneously (e.g., physics + chemistry for interdisciplinary questions).

- [ ] **8b. Pack versioning with diff** — Show what changed between pack versions (new articles, updated entities, eval score changes).

- [ ] **8c. Incremental pack updates** — Update a pack's graph without rebuilding from scratch (use `wikigr update` internally).

- [ ] **8d. Pack dependencies** — A pack can declare dependencies on other packs (e.g., quantum-computing-expert depends on physics-expert).

- [ ] **8e. Custom retrieval strategies** — Allow `kg_config.json` to specify retrieval strategies (vector-only, graph-only, hybrid weights, reranking).

- [ ] **8f. Pack quality monitoring** — Track eval scores over time as packs are updated; alert on quality regression.

- [ ] **8g. Automatic question generation** — Use LLM to automatically generate evaluation questions from pack content, reducing manual curation effort.

---

## Technical Debt

| Item | Location | Severity | Notes |
|------|----------|----------|-------|
| `_retrieve_context()` is a stub | `wikigr/packs/eval/baselines.py:165` | HIGH | Must wire to real KG Agent for meaningful evaluation |
| Baseline evaluators return mocked responses | `wikigr/packs/eval/baselines.py` | HIGH | Must call real Anthropic API for actual evaluation |
| `pack create` doesn't build graph | `wikigr/cli.py:cmd_pack_create()` | HIGH | Only creates scaffold; doesn't invoke `wikigr create` |
| Pack tests not in CI | `.github/workflows/ci.yml` | MEDIUM | 175 tests only run locally |
| `anthropic` SDK not in dependencies | `pyproject.toml` | MEDIUM | Eval framework imports it but it's not listed |
| BFS queue uses `list.pop(0)` | `wikigr/cli.py:206, 399` | LOW | O(n) dequeue; should use `collections.deque` |
| Duplicate BFS logic | `wikigr/cli.py` | LOW | `_create_from_urls` and `_update_from_urls` share ~50 lines |
| `PHASE4_DISTRIBUTION.md` in repo | `wikigr/packs/` | LOW | Temporal doc should be moved to logs |
| `SECURITY_FIXES.md` in repo root | root | LOW | Sprint artifact; info already in commit messages |

---

## Key Files Reference

| File | Purpose |
|------|---------|
| `wikigr/cli.py` | CLI entry point with all commands (1,245 LOC) |
| `wikigr/packs/__init__.py` | Pack system public API (84 LOC, 34 exports) |
| `wikigr/packs/manifest.py` | Manifest data models and I/O (233 LOC) |
| `wikigr/packs/eval/baselines.py` | 3-baseline evaluators (198 LOC) |
| `wikigr/packs/eval/runner.py` | Evaluation orchestration (118 LOC) |
| `wikigr/packs/eval/metrics.py` | Quality metrics calculation (166 LOC) |
| `wikigr/packs/skill_template.py` | Claude Code skill.md generation (119 LOC) |
| `wikigr/packs/installer.py` | Pack installation from file/URL (128 LOC) |
| `wikigr/packs/distribution.py` | Pack archiving (.tar.gz) (143 LOC) |
| `wikigr/agent/kg_agent.py` | KG Agent with hybrid retrieval (1,140 LOC) |
| `bootstrap/src/sources/web.py` | Web content source with SSRF protection (350 LOC) |
| `bootstrap/src/sources/base.py` | ContentSource protocol (76 LOC) |
| `bootstrap/src/expansion/processor.py` | Article processor pipeline (490 LOC) |
| `bootstrap/schema/ryugraph_schema.py` | Complete Kuzu schema |
| `docs/design/knowledge-packs.md` | Design specification (1,547 LOC) |

---

## Metrics

| Metric | Value |
|--------|-------|
| Total Python files | ~50 |
| Total test files | 28 |
| Tests passing | 262 (87 bootstrap + 175 pack) |
| PRs merged | 108 (106 original + #138 + #140) |
| Issues closed | 139 (all resolved) |
| CI jobs | 8 (main pipeline) + hygiene checks |
| Documentation files | 12 |
| CLI commands | 11 (3 main + 8 pack) |

---

## Summary

The platform is architecturally complete. The immediate priority is **proving value by creating a real knowledge pack, running a real evaluation with actual API calls, and demonstrating measurable improvement over baseline agent capabilities**. Everything else follows from that proof point.

**Comment by rysweet** [comment]
Superseded by #201 which tracks the complete system state.
